src/dockerflow/checks/registry.py

```diff
@@ -0,0 +1,201 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, you can obtain one at http://mozilla.org/MPL/2.0/.
+import asyncio
+import functools
+import inspect
+import logging
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, Iterable, List, Optional, Tuple
+
+from .messages import CheckMessage, level_to_text
+
+logger = logging.getLogger(__name__)
+
+CheckFn = Callable[..., List[CheckMessage]]
+
+_REGISTERED_CHECKS = {}
+
+
+def _iscoroutinefunction_or_partial(obj):
+    """
+    Determine if the provided object is a coroutine function or a partial function
+    that wraps a coroutine function.
+
+    This function should be removed when we drop support for Python 3.7, as this is
+    handled directly by `inspect.iscoroutinefunction` in Python 3.8.
+    """
+    while isinstance(obj, functools.partial):
+        obj = obj.func
+    return inspect.iscoroutinefunction(obj)
+
+
+def register(func=None, name=None):
+    """
+    TODO: This docstring
+    """
+    if func is None:
+        return functools.partial(register, name=name)
+
+    if name is None:
+        name = func.__name__
+
+    logger.debug("Registered Dockerflow check %s", name)
+
+    if _iscoroutinefunction_or_partial(func):
+
+        @functools.wraps(func)
+        async def decorated_function_asyc(*args, **kwargs):
+            logger.debug("Called Dockerflow check %s", name)
+            return await func(*args, **kwargs)
+
+        _REGISTERED_CHECKS[name] = decorated_function_asyc
+        return decorated_function_asyc
+    else:
+
+        @functools.wraps(func)
+        def decorated_function(*args, **kwargs):
+            logger.debug("Called Dockerflow check %s", name)
+            return func(*args, **kwargs)
+
+        _REGISTERED_CHECKS[name] = decorated_function
+        return decorated_function
+
+
+def init_check(check, obj):
+    """
+    Adds a given check callback with the provided object to the list
+    of checks. Useful for built-ins but also advanced custom checks.
+    """
+    logger.debug("Adding extension check %s" % check.__name__)
```

Comment:
Ahh I suppose it would. I'll put in back 👍🏻 These are great questions. I had ones that were similar. The `init_check` / "extension check" vocabulary existed before this PR -- it was a pattern for [Flask](https://github.com/mozilla-services/python-dockerflow/blob/0523a4efe2e8fec702adbe6bffcd42a550f47f5f/src/dockerflow/flask/app.py#L149) and [Sanic](https://github.com/mozilla-services/python-dockerflow/blob/0523a4efe2e8fec702adbe6bffcd42a550f47f5f/src/dockerflow/sanic/app.py#L116). I don't actually see any traces of it in Django 🤔.

It does seem like it could be reworked. I'll give it a try in this PR, but if things get too crazy, maybe it deserves its own.

Reply:
Oops you're right, misread the diff at the moment of writing the comment :)
Don't hesitate to open a follow-up if you think that's too much for this PR